flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java

```diff
@@ -134,10 +138,17 @@ public void cancel(InputChannelID receiverId) {
 		ctx.pipeline().fireUserEventTriggered(receiverId);
 	}
 
-	public void close() {
+	public void close() throws IOException {
 		if (ctx != null) {
 			ctx.channel().close();
 		}
+
+		LOG.info("Close all {} readers pending for close.", readersToClose.size());
```

Comment:
If the partition is released once it's transported, then the TM containing the partition will be released. It will cause a RemoteTransportException "Connection unexpectedly closed by remote task manager  ..." in CreditBasedPartitionRequestClientHandler's channelInactive() method, as the Netty connection is shared between TMs.

So I changed this to let writer to close the readers when closing the Netty connection. This is the smallest possible change I can find. Do you have any suggestions how we can handle this case better?

Reply:
Sorry for late response because of a little busy these days.

I agree with your above comments. So whether the task executor can be released is based on whether there are active channels in this executor. The task executor can only exit after all the tcp connections are closed gracefully.

In theory as long as the downstream received all the data from the network, then the upstream side can be released normally, no need to wait all the received data are processed completely by downstream side. But we have on existing ack mechanism to notify upstream side of this, so it is easy to rely on close request currently.  Based on downstream's consumption to release upstream's resource, it may get extra benefits in failover scenarios in future for persistent output files in upstream side, because the upstream do not need to restart during consumption exception in downstream side.

But I just a little wonder it might bring potential effects in future via close request. For example, if there are 10 downstream tasks reuse the same tcp connection, and 9 tasks are finished earlier and only one tail task delay long to finish, then all the 10 partition views must be released together until the last downstream task finished. Although it might be no bad effects for delay releasing partition views currently.

I would continue reviewing other parts of this PR and it may take some days on my side. :)

Reply:
Thanks Zhijiang. Agree with you that TM should only exit after all connections are closed. 

Regarding the side effect that 1 tail tasks out of 10 tasks will delay releasing partitions, I think this may not be bad since the resource is managed at TM level. If a TM cannot be released, it doesn't make too much difference whether it has 1 partition or 10 partitions. Please correct me if anything wrong.

Considering all existing limitations, I think the resource management could be further improved via external shuffle service that you've proposed. This PR may focus on solving current issue. Looking forward to your further comments. :)